lightning/src/ln/channelmanager.rs

```diff
@@ -3551,12 +3548,20 @@ where
 								}
 
 								macro_rules! check_total_value {
-									($payment_data: expr, $payment_preimage: expr) => {{
+									($payment_secret: expr, $payment_preimage: expr) => {{
 										let mut payment_claimable_generated = false;
-										let purpose = || {
+										let is_keysend = match claimable_htlc.onion_payload {
+											OnionPayload::Spontaneous(_) => true,
+											OnionPayload::Invoice { .. } => false,
+										};
+										let purpose = if is_keysend {
+											events::PaymentPurpose::SpontaneousPayment(
+												$payment_preimage.expect("Should never call check_total_value with keysend payment but no preimage")
+											)
+										} else {
 											events::PaymentPurpose::InvoicePayment {
 												payment_preimage: $payment_preimage,
-												payment_secret: $payment_data.payment_secret,
+												payment_secret: $payment_secret.expect("Should never call check_total_value with non-keysend payment but no payment secret"),
```

Comment:
Mm good point, realized I could just pass in `purpose` to `check_total_value` and avoid these unwraps altogether, will push these changes soon

Reply:
Okay, made this change, as well as made `is_keysend` based on this `purpose` that is passed into `check_total_value`, and both of those match the `claimable_htlc.onion_payload` based on where they're called

Also fixed an additional ambiguity ([here](https://github.com/lightningdevkit/rust-lightning/pull/2156/commits/250ee32779d19f7c2426ffd89e102610ffbcfbd1#diff-645d12c7269d09caab57d2e0c1fd34f672ee259994c86ad4bad7f4da9183671cL3691-R3689)) - for the legacy path where we'd store inbound payments, it looked at `payment_data.is_none()` to check if the payment was keysend; added a check on the onion payload since with mpp keysend `payment_data` could be `Some`, and in this path we should only be handling invoice payments